### PR TITLE
Various improvements to Kokkos version

### PR DESCRIPTION
**Pull request by makortel**
The first three commits are to bring the behavior of the Kokkos version closer to `cudacompat` (and `cuda`). The last commit is for cleaner code (that may become important for future architectures).

### CODE REVIEW

File: src/kokkos/KokkosCore/atomic.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#ifndef KokkosCore_atomic_h
+#define KokkosCore_atomic_h
+
+#include <Kokkos_Core.hpp>
+
+// Abstractions to disable atomicity of atomics for serial backend
+namespace cms::kokkos {
+  template <typename T>
+  KOKKOS_FORCEINLINE_FUNCTION T atomic_fetch_add(T* dest, const T val) {
```
Comment (makortel) [inline src/kokkos/KokkosCore/atomic.h:9]: Note to self that this is fragile at best, and these functions should be (e.g.) templated on the execution space to guarantee different symbols. Probably the `FORCEINLINE` has saved for trouble in the tests so far.

**Comment by fwyzard** [comment]
I guess these are needed because Kokkos does not provide no-op atomics for
the serial case.
Is it because there is no distinction of per-block, per-grid and per-device
atomics ?

**Comment by makortel** [comment]
> I guess these are needed because Kokkos does not provide no-op atomics for the serial case. 

Correct. Avoiding real atomics for the serial case was suggested by Kokkos developers as a step towards closing the `cudacompat` performance (real impact still to be quantified).

> Is it because there is no distinction of per-block, per-grid and per-device atomics ?

Good question. The API currently is, AFAICT, such that there is no distinction between different levels of atomicity guarantees, but we should find out if they have any plans for such.

File: src/kokkos/KokkosCore/atomic.h (hunk 1 above)
Comment (makortel) [reply src/kokkos/KokkosCore/atomic.h:15]: For now went with a simpler solution that the bypassing of atomics is an option available when only the Serial backend is enabled. This should be sufficient to demonstrate the impact on performance, after which some better solution can be figured out.

**Pull request merged**